extensions/android-paging3/src/main/java/com/squareup/sqldelight/android/paging3/OffsetQueryPagingSource.kt

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sqldelight.android.paging3
+
+import androidx.paging.PagingState
+import com.squareup.sqldelight.Query
+import com.squareup.sqldelight.Transacter
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+
+internal class OffsetQueryPagingSource<RowType : Any>(
+  private val queryProvider: (limit: Long, offset: Long) -> Query<RowType>,
+  private val countQuery: Query<Long>,
+  private val transacter: Transacter,
+  private val dispatcher: CoroutineDispatcher,
+) : QueryPagingSource<Int, RowType>() {
+
+  override suspend fun load(
+    params: LoadParams<Int>
+  ): LoadResult<Int, RowType> = withContext(dispatcher) {
+    try {
+      val key = params.key?.toLong() ?: 0L
+      transacter.transactionWithResult {
+        val count = countQuery.executeAsOne()
+        if (isOutOfBounds(key, params.loadSize, count)) throw IndexOutOfBoundsException()
+
+        val data = queryProvider(params.loadSize.toLong(), params.loadSize * key)
```

Comment:
Is there a reason why this is page-keyed (offset == key * pageSize I'm guessing here?) instead of position-keyed (offset == key)?

If you do the latter, then you can make `getRefreshKey` independent from loaded pages and only depend on `anchorPosition` so you can support jumping (fast-scroll) behavior, which let's you skip loading pages one at a time and calls refresh to restart loading from the middle of the list by using key returned from `getRefreshKey` with `anchorPosition != null`.

See: https://developer.android.com/reference/kotlin/androidx/paging/PagingSource#jumpingsupported

Reply:
Could you please explain a little further what you mean? As far as I'm aware, this implementation does support jumping. [source](https://github.com/cashapp/sqldelight/pull/2197/commits/95e3fcecd540faee8bc6b82a91cc914eefb93fba#diff-47e8065d8de3a6d1d692aab8977e23a181eb269117a76492526f2002b37b1ca3R139)

Is there something I'm missing? I do realize I should probably override `jumpingSupported`.

Reply:
That test looks like it just calls Refresh with the loadKey filled in, so I'm not sure it fully covers jumping.

For context - jumping is the fast-scroll behavior where you want to restart loading from the middle of the list to avoid appending until you get to user's scroll position (useful for super long lists or expensive loads). Paging handles this by sending another `PagingSource.load()` call with `LoadParams.Refresh` where key is set by calling `getRefreshKey()` with non-null anchorPosition and then completely replaces the existing list and resumes sequential loading from there.

The issue is that this is typically going to be a scroll position in placeholders, which is far away from any loaded pages, so since `getRefreshKey()` works by accessing closest loaded page and incrementing / decrementing, it will often provide the wrong key on a jump.

Actually I notice now that placeholders isn't filled in `LoadResult.Page()`, so that should probably happen too since you already make the query to get the total count. You can do this by filling in `itemsBefore` and `itemsAfter` in `LoadResult.Page` constructor.

Both behaviors are opt-in on use-site in `PagingConfig` as well. Developer can control whether placeholders are disabled and whether they want jumping behavior enabled, doing it in `PagingSource` just gives them the option to enable those.

Reply:
Thanks for this review! Super helpful. I just pushed some updates